src/Common/tests/System/Collections/ICollection.Generic.Tests.cs

```diff
@@ -0,0 +1,622 @@
+    /// ICollection interface
+    /// </summary>
+    public abstract class ICollection_Generic_Tests<T> : IEnumerable_Generic_Tests<T>
+    {
+        #region ICollection<T> Helper Methods
+
+        /// <summary>
+        /// Creates an instance of an ICollection{T} that can be used for testing.
+        /// </summary>
+        /// <returns>An instance of an ICollection{T} that can be used for testing.</returns>
+        protected abstract ICollection<T> GenericICollectionFactory();
+
+        /// <summary>
+        /// Creates an instance of an ICollection{T} that can be used for testing.
+        /// </summary>
+        /// <param name="count">The number of unique items that the returned ICollection{T} contains.</param>
+        /// <returns>An instance of an ICollection{T} that can be used for testing.</returns>
+        protected virtual ICollection<T> GenericICollectionFactory(int count)
+        {
+            ICollection<T> collection = GenericICollectionFactory();
+            AddToCollection(collection, count);
+            return collection;
+        }
+
+        protected virtual bool DuplicateValuesAllowed { get { return true; } }
+        protected virtual bool DefaultValueWhenNotAllowed_Throws { get { return true; } }
+        protected virtual bool IsReadOnly { get { return false; } }
+        protected virtual bool DefaultValueAllowed { get { return true; } }
+        protected virtual IEnumerable<T> InvalidValues { get { return Array.Empty<T>(); } }
+        protected virtual void AddToCollection(ICollection<T> collection, int numberOfItemsToAdd)
+        {
+            int seed = 9600;
+            IEqualityComparer<T> comparer = GetIEqualityComparer();
+            while (collection.Count < numberOfItemsToAdd)
+            {
+                T toAdd = TFactory(seed++);
+                while (collection.Contains(toAdd, comparer) || InvalidValues.Contains(toAdd, comparer))
+                    toAdd = TFactory(seed++);
+                collection.Add(toAdd);
+            }
+        }
+        
+        #endregion
+
+        #region IEnumerable<T> Helper Methods
+
+        protected override IEnumerable<T> GenericIEnumerableFactory(int count)
+        {
+            return GenericICollectionFactory(count);
+        }
+
+        protected override int WaysToModify { get { return 3; } } // add, remove, clear
+
+        protected override void ModifyEnumerable(IEnumerable<T> enumerable, int enumerationCode)
+        {
+            ICollection<T> casted = (ICollection<T>)enumerable;
+            switch (enumerationCode)
+            {
+                case 0: // Add
+                    casted.Add(TFactory(12));
+                    break;
+                case 1: // Remove
+                    if (casted.Count() > 0)
+                        casted.Remove(casted.ElementAt(0));
+                    else
+                        casted.Add(TFactory(12));
```

Comment:
This might be correct, but reading this code is confusing to me... the enumerationCode dictates a removal, but if there's nothing in the collection, we add to it?